go-controller/pkg/node/gateway_iptables.go

```diff
@@ -0,0 +1,324 @@
+		}
+		exists, err := ipt.Exists(r.table, r.chain, r.args...)
+		if !exists && err == nil {
+			err = ipt.Insert(r.table, r.chain, 1, r.args...)
+		}
+		if err != nil {
+			return fmt.Errorf("failed to add iptables %s/%s rule %q: %v",
+				r.table, r.chain, strings.Join(r.args, " "), err)
+		}
+	}
+	return nil
+}
+
+func delIptRules(rules []iptRule) error {
+	for _, r := range rules {
+		ipt, _ := util.GetIPTablesHelper(r.protocol)
+		err := ipt.Delete(r.table, r.chain, r.args...)
+		if err != nil {
+			return fmt.Errorf("failed to delete iptables %s/%s rule %q: %v", r.table, r.chain,
+				strings.Join(r.args, " "), err)
+		}
+	}
+	return nil
+}
+
+func ensureChain(ipt util.IPTablesHelper, table, chain string) error {
+	chains, err := ipt.ListChains(table)
+	if err != nil {
+		return fmt.Errorf("failed to list iptables chains: %v", err)
+	}
+	for _, ch := range chains {
+		if ch == chain {
+			return nil
+		}
+	}
+	return ipt.NewChain(table, chain)
+}
+
+func initSharedGatewayIPTables() error {
+	rules := make([]iptRule, 0)
+	for _, chain := range []string{iptableNodePortChain, iptableExternalIPChain} {
+		for _, proto := range []iptables.Protocol{iptables.ProtocolIPv4, iptables.ProtocolIPv6} {
+			ipt, err := util.GetIPTablesHelper(proto)
+			if err != nil {
+				return err
+			}
+			_ = ipt.ClearChain("nat", chain)
+			_ = ipt.ClearChain("filter", chain)
+
+			rules = append(rules, iptRule{
+				table:    "nat",
+				chain:    "OUTPUT",
+				args:     []string{"-j", chain},
+				protocol: proto,
+			})
+			rules = append(rules, iptRule{
+				table:    "nat",
+				chain:    "PREROUTING",
+				args:     []string{"-j", chain},
+				protocol: proto,
+			})
+			rules = append(rules, iptRule{
+				table:    "filter",
+				chain:    "OUTPUT",
+				args:     []string{"-j", chain},
+				protocol: proto,
+			})
+			rules = append(rules, iptRule{
+				table:    "filter",
+				chain:    "FORWARD",
+				args:     []string{"-j", chain},
+				protocol: proto,
+			})
+		}
+	}
+	if err := addIptRules(rules); err != nil {
+		return fmt.Errorf("failed to add iptable rules %v: %v", rules, err)
+	}
+	return nil
+}
+
+func cleanupSharedGatewayIPTChains() {
+	for _, chain := range []string{iptableNodePortChain, iptableExternalIPChain} {
+		for _, proto := range []iptables.Protocol{iptables.ProtocolIPv4, iptables.ProtocolIPv6} {
+			ipt, err := util.GetIPTablesHelper(proto)
+			if err != nil {
+				return
+			}
+			_ = ipt.ClearChain("nat", chain)
+			_ = ipt.ClearChain("filter", chain)
+			_ = ipt.DeleteChain("nat", chain)
+			_ = ipt.DeleteChain("filter", chain)
+		}
+	}
+}
+
+func initLocalGatewayIPTables() error {
+	rules := make([]iptRule, 0)
+	for _, chain := range []string{iptableNodePortChain, iptableExternalIPChain} {
+		for _, proto := range []iptables.Protocol{iptables.ProtocolIPv4, iptables.ProtocolIPv6} {
+			ipt, _ := util.GetIPTablesHelper(proto)
+			if err := ipt.NewChain("nat", chain); err != nil {
```

Comment:
`ensureChain` has the same end-result as `NewChain`, but less efficiently. From the doc: `NewChain creates a new chain in the specified table. If the chain already exists, it will result in an error.`. I will remove `ensureChain`. 

The difference between them are the chains they create: local gateway mode does not create "filter" "OUTPUT" (I don't know why that is exactly, I just took what was existing)